File: test/benchmark/beng/js/pidigits.js
// BENG Benchmark: pi-digits (Node.js reference)
// Unbounded spigot algorithm (Gibbons 2004) using BigInt

const NUM_DIGITS = parseInt(process.argv[2] || "30");

function idiv(a, b) {
    // Floor division for BigInts
    const d = a / b;
    // JS BigInt division truncates toward zero; adjust for negative
    if ((a < 0n) !== (b < 0n) && d * b !== a) return d - 1n;
    return d;
}

let q = 1n, r = 0n, s = 0n, t = 1n, k = 0n;
let i = 0;
let digits = '';

while (i < NUM_DIGITS) {
    k = k + 1n;
    const k2 = k * 2n + 1n;

    // compose: multiply LFT by next term
    const new_q = q * k;
    const new_r = (2n * q + r) * k2;
    const new_s = s * k;
    const new_t = (2n * s + t) * k2;
    q = new_q;
    r = new_r;
    s = new_s;
    t = new_t;

    // can we extract a digit?
    if (q <= r) {
        const fd3 = idiv(3n * q + r, 3n * s + t);
        const fd4 = idiv(4n * q + r, 4n * s + t);
        if (fd3 === fd4) {
            digits += fd3.toString();
            i++;

            if (i % 10 === 0) {
                console.log(`${digits}\t:${i}`);
                digits = '';
            }

            // reduce: eliminate the extracted digit
            r = (r - fd3 * t) * 10n;
            q = q * 10n;
        }
    }
}

if (digits.length > 0) {
    const pad = ' '.repeat(10 - digits.length);
    console.log(`${digits}${pad}\t:${i}`);
}

File: test/benchmark/beng/python/pidigits.py
#!/usr/bin/env python3
# BENG Benchmark: pi-digits (Python)
# Unbounded spigot algorithm (Gibbons 2004)
import sys
import time

NUM_DIGITS = int(sys.argv[1]) if len(sys.argv) > 1 else 30

t0 = time.perf_counter_ns()
q, r, s, t, k = 1, 0, 0, 1, 0
i = 0
digits = ''

while i < NUM_DIGITS:
    k += 1
    k2 = k * 2 + 1

    new_q = q * k
    new_r = (2 * q + r) * k2
    new_s = s * k
    new_t = (2 * s + t) * k2
    q, r, s, t = new_q, new_r, new_s, new_t

    if q <= r:
        fd3 = (3 * q + r) // (3 * s + t)
        fd4 = (4 * q + r) // (4 * s + t)
        if fd3 == fd4:
            digits += str(fd3)
            i += 1

            if i % 10 == 0:
                print(f"{digits}\t:{i}")
                digits = ''

            r = (r - fd3 * t) * 10
            q = q * 10

if digits:
    pad = ' ' * (10 - len(digits))
    print(f"{digits}{pad}\t:{i}")

t1 = time.perf_counter_ns()
print(f"__TIMING__:{(t1 - t0) / 1e6:.3f}")
